src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs

```diff
@@ -15,6 +15,7 @@ namespace System.Text.RegularExpressions
     /// <summary>
     /// RegexCompiler translates a block of RegexCode to MSIL, and creates a subclass of the RegexRunner type.
     /// </summary>
+    [RequiresDynamicCode("The native code for Regex compilation might not be available at runtime.")]
```

Comment:
Before we can place it on a class we need to:

* Get the updated attribute through an API review
* Make sure the Roslyn analyzer can actually support this. E.g. `RequiresAssemblyFiles` cannot be placed on a class, but it can be placed on a property. We're currently not very consistent on the applicability of these. Maybe we don't need to be? (Allowing `RequiresUnreferencedCode`/`RequiresDynamicCode` on properties is going to be bunch of work, for example)
* Make sure the NativeAOT compiler supports this.

If this pull request takes a dependency on this, it will be blocked until all of the above is validated/happens.

Reply:
I think the main value of this PR is that we are dogfooding the system we are putting in place and finding all the holes before customers will find them.

no public APIs are going to get these annotations. So really this is just for our internal code.

The alternatives to waiting would be:

1. Add attributes to most (all?) the methods in this class until we can support the class level attribute.

2.Suppress the warnings at the low level places like the original suggestion.

I could live with 1 for a while as long as we ensured we supported class level in .NET 7. We would also use that in JSON and DependencyInjection, since both of those libraries have whole classes just for generating IL.

Reply:
I like 1 - unless @LakshanF has a different opinion, let's create an issue tracking the 3 things above (API review, validate/implement Roslyn analyzer/add a test for it, validate/implement in NativeAOT and add a test for it, and undo workaround from this pull request). For this pull request, add RDC on the individual methods instead of on the class.

Reply:
Option 1 as mentioned above would require some churn - the warnings come from static read only fields that would require a longer workarounds that will need to be spilled into methods in the class - that we expect to be temporary anyway. 

I like @eerhardt thinking of this PR as flushing out initial issues with the annotation process and created an [API request](https://github.com/dotnet/runtime/issues/67368) for enhancing the attribute. I would like to explore that path (including compiler changes ) in more detail first since its the solution we eventually likely go with before going for the temporary long workaround with option 1.